punchbowl/levelq/pca.py

```diff
@@ -11,114 +18,203 @@
+    for i, mask in enumerate(blend_masks):
+        # We mark the images that don't have any planets in the quarter we're filtering for (since those can
+        # contaminate the PCA components)
+        no_bodies_in_quarter = np.all(bodies_in_quarter[:, :, i] == False, axis=1) # noqa: E712
+        images_to_fit = _all_files_to_fit[no_bodies_in_quarter * ~to_subtract_filter]
+        tag = f"stride {stride}, quarter {i+1}"
+        logger.info(f"Starting to filter {tag}, fitting {len(images_to_fit)} images")
+        filtered_by_quarter = run_pca_filtering(images_to_subtract, images_to_fit, n_components, med_filt, tag,
+                                                masked_region, logger)
+        if i == 0:
+            final_reconstruction = mask * filtered_by_quarter
+        else:
+            final_reconstruction += mask * filtered_by_quarter
+
+    return subtracted_cube_indices, final_reconstruction
+
+
+def run_pca_filtering(images_to_subtract: np.ndarray, images_to_fit: np.ndarray, n_components: int,
+                      med_filt: int, tag: str, masked_region: np.ndarray, logger: logging.Logger) -> np.ndarray:
     """Run PCA filtering."""
-    logger = get_run_logger()
-    with threadpool_limits(30):
-        pca = PCA(n_components=n_components)
-        pca.fit(all_files_to_fit.reshape((len(all_files_to_fit), -1)))
-        logger.info("Fitting finished")
+    pca = PCA(n_components=n_components)
+    # The image array has to be re-shaped into (n_images, n_pixels). (i.e., PCA wants 1D vectors, not 2D images)
+    pca.fit(images_to_fit[:, masked_region])
+    logger.info(f"Fitting finished for {tag}")
 
-        transformed = pca.transform(input_cube.data.reshape((1, -1)))
+    transformed = pca.transform(images_to_subtract[:, masked_region])
 
-        if med_filt:
-            for i in range(len(pca.components_)):
-                comp = pca.components_[i].reshape(input_cube.data.shape)
-                comp = scipy.signal.medfilt2d(comp, med_filt)
-                pca.components_[i] = comp.ravel()
-            logger.info("Median smoothing finished")
+    if med_filt:
+        for i in range(len(pca.components_)):
+            comp = np.zeros(images_to_fit.shape[1:], dtype=pca.components_[i].dtype)
+            comp[masked_region] = pca.components_[i]
+            comp = scipy.signal.medfilt2d(comp, med_filt)
+            pca.components_[i] = comp[masked_region]
+        logger.info(f"Median smoothing finished for {tag}")
 
-        reconstructed = pca.inverse_transform(transformed).reshape(input_cube.data.shape)
-        return input_cube.data - reconstructed
+    reconstructed = np.zeros_like(images_to_subtract)
+    reconstructed[:, masked_region] = pca.inverse_transform(transformed)
+    return images_to_subtract - reconstructed
 
 
-def find_bodies_in_image(frame: str | NDCube | WCS) -> list:
+def find_bodies_in_image(frame: str | NDCube | tuple[NormalizedMetadata, WCS], extra_padding: int = 0) -> list:
```

Comment:
This seems more general purpose. We might move this `find_bodies_in_image` function to somewhere like `punchbowl.data.wcs`

Reply:
Maybe, but this specifically marks which quarters of the image the planets are in, which feels less general.